### ISSUE TITLE
Skip unnecessary RegexNode reductions in NonBacktracking mode

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#62681]
As part of parsing a regex,  we perform various transformations on the tree, many of which are focused on creating a tree that will yield more efficient matching with a backtracking engine,  e.g. we'll transform "this|that" into "th(?:is|at)". But many of these transformations don't affect a finite automata built from this tree,  e.g. "this|that" and "th(?:is|at)" yield identical DFAs. Thus, in NonBacktracking mode,  we should avoid spending any cycles doing these transformations that won't make a difference.

**Comment by danmoseley** [comment]
Does this show up in a profile such that it's worth the special casing?

**Comment by stephentoub** [comment]
> Does this show up in a profile such that it's worth the special casing?

Yes.  And we also already special-case out of necessity, e.g. the analysis we do to automatically apply atomicity are disabled because atomic groups aren't supported in the DFA-based implementation today.
